### PR TITLE
RL incremental refactor

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

- Model
    - Change a set of abstract methods to concrete methods in `AbsNet` to decrease user effort when writing RL instances. Users could overwrite these methods when necessary. In most cases, using default ones is enough.
    - Add algorithm specific nets (e.g., `DiscreteACBasedNet`). These intermediate classes block unuseful methods so users could concentrate on the methods they should implement.
    - Refactor interfaces of `AbsNet`.
- Policy
    - Refactor interfaces of `RLPolicy`.
- Trainer
    - Change Discrete AC & Discrete PPO to general ones (both discrete & continuous)
    - Add SAC
    - Fix several minor bugs
- Workflow
    - Add `MIN_N_SAMPLE` in the main workflow. `EnvSampler` will keep collecting experiences until we have at least `MIN_N_SAMPLE` records.
    - Refine experience recording logic.

PPO, AC, SAC, DDPG have been tested on gym env. Gym simulator is on another branch that is based on the current branch.

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [ ] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/training/algorithms/ddpg.py [hunk 1]
```diff
@@ -40,6 +37,7 @@ class DDPGParams(TrainerParams):
     q_value_loss_cls: Callable = None
     soft_update_coef: float = 1.0
     random_overwrite: bool = False
+    n_start_train: int = 0
```
Comment (Jinyu-W) [inline maro/rl/training/algorithms/ddpg.py:40]: renamed to sth. like `min_num_to_trigger_training` ?

Comment (Jinyu-W) [reply maro/rl/training/algorithms/ddpg.py:40]: What's the relationship between this and `min_n_sample` in main.py? If they are for the same purpose, can we remove this here?

File: maro/rl/rollout/env_sampler.py [hunk 2]
```diff
@@ -416,7 +417,7 @@ def sample(self, policy_state: Optional[Dict[str, object]] = None, num_steps: Op
         return {
             "end_of_episode": not self._agent_state_dict,
             "experiences": [experiences],
-            "info": [self._info],
+            "info": [deepcopy(self._info)],
```
Comment (Jinyu-W) [inline maro/rl/rollout/env_sampler.py:420]: add `TODO` for for potential performance issue

File: maro/rl/training/algorithms/sac.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+
+class SoftActorCriticOps(AbsTrainOps):
+    def __init__(
+        self,
+        name: str,
+        policy_creator: Callable[[], RLPolicy],
+        get_q_critic_net_func: Callable[[], QNet],
+        parallelism: int = 1,
+        *,
```
Comment (Jinyu-W) [inline maro/rl/training/algorithms/sac.py:45]: Can we add "**args" to the base class to fix the potential issue without `*`?

Comment (lihuoran) [reply maro/rl/training/algorithms/sac.py:45]: Let's just rewrite this part in "normal" fashion.

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**